File: ios/Touliao/Core/Storage/DraftStore.swift
import Foundation

struct DraftOwner: Equatable {
    let server: String
    let accountId: String
    let identityEpoch: UInt64
    let generation: UInt64
}

/// Text drafts only; legacy vxin_draft_<conversation> rows have no proven owner.
/// Keep them quarantined without assigning or deleting another account's data.
final class DraftStore {
    static let shared = DraftStore(defaults: .standard, environment: {
        (ServerConfig.shared.baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/")), KeychainStore.shared.snapshot().identityEpoch)
    })
    private let defaults: UserDefaults
    private let environment: (() -> (String, UInt64))?
    private let lock = NSRecursiveLock()
    private var active: DraftOwner?
    private var generation: UInt64 = 0
    init(defaults: UserDefaults = .standard, environment: (() -> (String, UInt64))? = nil) {
        self.defaults = defaults
        self.environment = environment
    }
    @discardableResult func activate(server: String, accountId: String, identityEpoch: UInt64) -> DraftOwner {
        lock.lock(); defer { lock.unlock() }
        let normalized = server.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        if let active, active.server == normalized, active.accountId == accountId, active.identityEpoch == identityEpoch { return active }
        generation += 1
        let owner = DraftOwner(server: normalized, accountId: accountId, identityEpoch: identityEpoch, generation: generation)
        active = owner
        return owner
    }
    func invalidate() { lock.lock(); defer { lock.unlock() }; active = nil; generation += 1 }
    func capture() -> DraftOwner? { lock.lock(); defer { lock.unlock() }; return active }
    private func current(_ owner: DraftOwner?, environment now: (String, UInt64)?) -> Bool {
        guard let owner, owner == active, !owner.accountId.isEmpty, !owner.server.isEmpty else { return false }
        if let now { return now.0 == owner.server && now.1 == owner.identityEpoch }
        return true
    }
    private func key(_ owner: DraftOwner, _ conversation: String) -> String {
        "vxin_draft_v2:" + [owner.server, owner.accountId, conversation].map { "\($0.utf8.count):\($0)" }.joined()
    }
    func get(_ conversationId: String) -> String { get(conversationId, owner: capture()) }
    func get(_ conversationId: String, owner: DraftOwner?) -> String {
        let now = environment?() // Credential lock precedes the draft lock.
        lock.lock(); defer { lock.unlock() }
        guard !conversationId.isEmpty, current(owner, environment: now), let owner else { return "" }
        return defaults.string(forKey: key(owner, conversationId)) ?? ""
    }
    func set(_ conversationId: String, _ text: String, owner: DraftOwner?) {
        let now = environment?()
        lock.lock(); defer { lock.unlock() }
        guard !conversationId.isEmpty, current(owner, environment: now), let owner else { return }
        let storageKey = key(owner, conversationId)
        if text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { defaults.removeObject(forKey: storageKey) }
        else { defaults.set(text, forKey: storageKey) }
    }
    func clear(_ conversationId: String, owner: DraftOwner?) { set(conversationId, "", owner: owner) }
}

File: android/app/src/main/java/com/touliao/app/core/storage/DraftStore.kt
package com.touliao.app.core.storage

import android.content.Context
import android.content.SharedPreferences
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject
import javax.inject.Singleton

data class DraftOwner(val server: String, val accountId: String, val identityEpoch: Long, val generation: Long)

/** Text drafts only. Unowned legacy keys remain quarantined and are never imported. */
@Singleton
class DraftStore internal constructor(
    private val prefs: SharedPreferences,
    private val environment: (() -> Pair<String, Long>)? = null,
) {
    @Inject constructor(@ApplicationContext context: Context, tokens: TokenStore, server: ServerConfig) : this(
        context.getSharedPreferences("vxin_drafts", Context.MODE_PRIVATE),
        { server.baseUrl.trimEnd('/') to tokens.snapshot().identityEpoch },
    )
    private var generation = 0L
    private var active: DraftOwner? = null

    @Synchronized fun activate(server: String, accountId: String, identityEpoch: Long): DraftOwner {
        val normalized = server.trimEnd('/')
        active?.let { if (it.server == normalized && it.accountId == accountId && it.identityEpoch == identityEpoch) return it }
        return DraftOwner(normalized, accountId, identityEpoch, ++generation).also { active = it }
    }
    @Synchronized fun invalidate() { active = null; generation++ }
    @Synchronized fun capture(): DraftOwner? = active
    private fun current(owner: DraftOwner?, env: Pair<String, Long>?): Boolean = owner != null && owner == active &&
        owner.accountId.isNotBlank() && owner.server.isNotBlank() &&
        (env?.let { it.first == owner.server && it.second == owner.identityEpoch } ?: true)
    // Length-prefix encoding prevents separator collisions in server/account/conversation IDs.
    private fun key(owner: DraftOwner, conversationId: String) = "v2:" +
        listOf(owner.server, owner.accountId, conversationId).joinToString("") { "${it.length}:$it" }
    fun get(conversationId: String, owner: DraftOwner? = capture()): String {
        // Read the credential lock before the draft lock, matching SessionManager.
        val env = environment?.invoke()
        return synchronized(this) {
            if (conversationId.isBlank() || !current(owner, env)) ""
            else prefs.getString(key(owner!!, conversationId), "").orEmpty()
        }
    }
    fun set(conversationId: String, text: String, owner: DraftOwner?) {
        val env = environment?.invoke()
        synchronized(this) {
            if (conversationId.isBlank() || !current(owner, env)) return
            val key = key(owner!!, conversationId)
            prefs.edit().apply { if (text.isBlank()) remove(key) else putString(key, text) }.apply()
        }
    }
    fun clear(conversationId: String, owner: DraftOwner?) = set(conversationId, "", owner)
}
